src/docfx/lib/git/FileCommitProvider.cs

```diff
@@ -193,145 +192,162 @@ public void Dispose()
+                    var author = git_commit_author(commit);
+                    var parentCount = git_commit_parentcount(commit);
+                    var parents = new git_oid[parentCount];
+                    for (var i = 0; i < parentCount; i++)
+                    {
+                        parents[i] = *git_commit_parent_id(commit, i);
+                    }
 
-                var item = new Commit(commitId, parents, *git_commit_tree_id(commit), gitCommit);
-                commitsBySha.Add(commitId.a, item);
-                commits.Add(item);
-                git_object_free(commit);
-            }
-            git_revwalk_free(walk);
+                    var gitCommit = new GitCommit(
+                        Marshal.PtrToStringUTF8(author->name) ?? "",
+                        Marshal.PtrToStringUTF8(author->email) ?? "",
+                        commitId.ToString(),
+                        new git_time { time = git_commit_time(commit), offset = git_commit_time_offset(commit) }.ToDateTimeOffset());
 
-            // build parent indices
-            Parallel.ForEach(commits, commit =>
-            {
-                for (var i = 0; i < commit.ParentShas.Length; i++)
-                {
-                    commit.Parents[i] = commitsBySha[commit.ParentShas[i].a];
+                    var treeId = *git_commit_tree_id(commit);
+                    var tree = new Tree { Id = treeId };
+                    var item = new Commit(commitId, parents, tree, gitCommit);
+
+                    commitsById.Add(commitId.a, item);
+                    commits.Add(item);
+                    git_object_free(commit);
                 }
-                commit.ParentShas = Array.Empty<git_oid>();
-            });
+                git_revwalk_free(walk);
 
-            return (commits, commitsBySha);
+                // build parent indices
+                Parallel.ForEach(commits, commit =>
+                {
+                    for (var i = 0; i < commit.ParentIds.Length; i++)
+                    {
+                        commit.Parents[i] = commitsById[commit.ParentIds[i].a];
+                    }
+                    commit.ParentIds = Array.Empty<git_oid>();
+                });
+
+                return (commits, commitsById);
+            }
         }
 
-        private long GetBlob(git_oid treeId, int[] pathSegments)
+        private long GetBlob(Tree tree, uint[] pathSegments)
         {
-            var blob = treeId;
+            var node = tree;
 
             for (var i = 0; i < pathSegments.Length; i++)
             {
-                var files = _trees.GetOrAdd(blob.a, _ => LoadTree(blob));
-                if (files is null || !files.TryGetValue(pathSegments[i], out blob))
+                var children = node.Children ?? LoadChildren(node);
+                if (ReferenceEquals(children, s_emptyTree) || !children.TryGetValue(pathSegments[i], out node))
```

Comment:
if `s_emptyTree` is only used for checking no children, why not just `children.Count == 0`?

Reply:
yeah, both are fine